Commands to Help Manage Resources and Remote Uploaded Files via LibCloud

JVickery-TBS opened the pull request:
feat(commands): added commands to list unlinked resources and uploads;

- Added list-missing-uploads command.
- Added list-unlinked-uploads command.

JVickery-TBS committed:
feat(commands): added commands to list unlinked resources and uploads;

- Added list-missing-uploads command.
- Added list-unlinked-uploads command.

JVickery-TBS committed:
feat(commands): added placeholder commands;

- Added placeholder fix-missing-uploads command.
- Added placeholder remove-unlinked-uploads command.

JVickery-TBS committed:
feat(commands): added delete command code;

- Added code to remove unlinked uploads.
- Removed fix-missing-uploads placeholder command.
- Small code improvements.

JVickery-TBS commented:
@wardi Okay this should be good for reviewing now!

wardi commented on ckanext/cloudstorage/cli.py:
```diff
@@ -164,6 +179,96 @@ def _fix_cors(args):
+def _get_unlinked_uploads(return_objects = False):
+    cs = CloudStorage()
+
+    resource_urls = []
```
use a `set()` here so that the `if upload.name not in resource_urls` below is faster

JVickery-TBS replied:
`resource_urls` now updated to a set.

JVickery-TBS replied:
@wardi can you confirm that the new set code is good? I compounded it into just the one var.

wardi commented on ckanext/cloudstorage/cli.py:
```diff
@@ -164,6 +179,96 @@ def _fix_cors(args):
+def _list_unlinked_uploads():
+    uploads_missing_resources = _get_unlinked_uploads()
+
+    if len(uploads_missing_resources):
```
```suggestion
    if uploads_missing_resources:
```
empty containers are False and non-empty lists are True in python

wardi commented on ckanext/cloudstorage/cli.py:
```diff
@@ -164,6 +179,96 @@ def _fix_cors(args):
+    upload_urls = []
+    uploads = cs.container.list_objects()
+    for upload in uploads:
+        upload_urls.append(upload.name)
```
```suggestion
    upload_urls = set(u.name for u in cs.container.list_objects())
```

wardi commented on ckanext/cloudstorage/cli.py:
```diff
@@ -164,6 +179,96 @@ def _fix_cors(args):
+        if url not in upload_urls:
+            resource_ids_missing_uploads.append(id)
+
+    if len(resource_ids_missing_uploads):
```
```suggestion
    if resource_ids_missing_uploads:
```

JVickery-TBS committed:
Update ckanext/cloudstorage/cli.py

Co-authored-by: Ian Ward <ian@excess.org>

JVickery-TBS committed:
Update ckanext/cloudstorage/cli.py

Co-authored-by: Ian Ward <ian@excess.org>

JVickery-TBS committed:
Update ckanext/cloudstorage/cli.py

Co-authored-by: Ian Ward <ian@excess.org>

JVickery-TBS committed:
feat(commands): changed list to set;

- Changed list to set.
- Made set fancy.

wardi commented on ckanext/cloudstorage/cli.py:
```diff
@@ -164,6 +179,91 @@ def _fix_cors(args):
+def _get_unlinked_uploads(return_objects = False):
+    cs = CloudStorage()
+
+    resource_urls = (os.path.join(
```
just need
```suggestion
    resource_urls = set(os.path.join(
```
otherwise you're creating a generator.

JVickery-TBS committed:
Update ckanext/cloudstorage/cli.py

Co-authored-by: Ian Ward <ian@excess.org>

JVickery-TBS committed:
feat(commands): csv output and list-uploads;

- Added `list-uploads` command.
- Added output option.
- Added csv file output.

JVickery-TBS commented:
@wardi okay there is now a `list-uploads` commands which outputs the linked uploads to resources.

There is now an `output` option to the command which is used by `list-uploads`, `list-unlinked-uploads`, and `list-missing-uploads` commands.

The headers and data for the csv outputs are:
* resource_id
* package_id
* organization_id
* resource_filename
* upload_url
* upload_file_size
* resource_created
* resource_last_modified

wardi commented:
Sounds great. looking forward to reports on our sites' cloudstorage containers.

JVickery-TBS committed:
fix(commands): syntax fix;

- Typo fix.

JVickery-TBS committed:
feat(commands): added human readable size;

- Added human readable size formats for space used/saved.
